File: frameworks/payload.core/src/main/java/com/qomolangma/frameworks/payload/core/ListElement.java
package com.qomolangma.frameworks.payload.core;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ListElement<D> {
    private final ConvertRules<D> conventions;

    @SafeVarargs
    public ListElement(ConvertRule<D>... conventions) {
        this.conventions = new ConvertRules<>(conventions);
    }

    @SafeVarargs
    public static <D> ListElement<D> list(ConvertRule<D>... conventions) {
        return new ListElement<>(conventions);
    }

    public List<Map<String, Object>> transform(List<D> objects) {
        return objects.stream().map(conventions::convert).collect(Collectors.toList());
    }
}

File: frameworks/payload.core/src/main/kotlin/com/qomolangma/frameworks/payload/core/ListElement.kt
package com.qomolangma.frameworks.payload.core

import java.util.stream.Collectors

class ListElement<D> constructor(vararg conventions: ConvertRule<D>) {
    private val conventions: ConvertRules<D>

    init {
        this.conventions = ConvertRules(*conventions)
    }

    fun transform(objects: List<D>): List<Map<String, Any>> {
        return objects.stream().map(conventions::convert).collect(Collectors.toList())
    }

    companion object {
        @JvmStatic
        fun <D> list(vararg conventions: ConvertRule<D>): ListElement<D> {
            return ListElement(*conventions)
        }
    }
}
